### PR TITLE
Wip/hughsie/elaptp

### PR DESCRIPTION
**Pull request by hughsie**

### CODE REVIEW

File: plugins/elantp/elantp.quirk [hunk 1]
```diff
@@ -6,18 +6,54 @@ GType = FuElantpHidDevice
...
+ElantpIcPageCount = 0x500
+
+[DeviceInstanceId=ELANTP\ICTYPE_13]
+ElantpIcPageCount = 0x800
+
+[DeviceInstanceId=ELANTP\ICTYPE_14]
+ElantpIcPageCount = 0x400
+
+[DeviceInstanceId=ELANTP\ICTYPE_15]
+ElantpIcPageCount = 0x400
```
Comment (jinglewu) [inline plugins/elantp/elantp.quirk:59]: ICTYPE_11 13 14 15 -> ElantpIapPassword = 0x1EA5

File: plugins/elantp/fu-elantp-hid-device.c [hunk 2]
```diff
@@ -215,6 +217,13 @@ fu_elantp_hid_device_setup (FuDevice *device, GError **error)
 	instance_id_ic_type = g_strdup_printf ("ELANTP\\ICTYPE_%02X", ic_type);
 	fu_device_add_instance_id (device, instance_id_ic_type);
```
Comment (jinglewu) [inline plugins/elantp/fu-elantp-hid-device.c:219]: fwver is mean iap_version??

File: plugins/elantp/fu-elantp-hid-device.c [hunk 3]
```diff
@@ -215,6 +217,13 @@ fu_elantp_hid_device_setup (FuDevice *device, GError **error)
...
 
+	/* get the page size */
+	if (fwver >= 2 && (ic_type == 0x14 || ic_type == 0x15)) {
+		self->fw_page_size = 512;
+	} else if (ic_type >= 0x10) {
```
Comment (jinglewu) [inline plugins/elantp/fu-elantp-hid-device.c:223]: iap_version >= 1 && ic_type >=10 (require iap_version >= 1)
if iap_version==0 then -> page_size = 64

File: plugins/elantp/fu-elantp-hid-device.c [hunk 4]
```diff
@@ -166,14 +166,18 @@ fu_elantp_hid_device_setup (FuDevice *device, GError **error)
...
 	fu_device_set_version (device, version);
 
-	/* get EAP firmware version */
+	/* get IAP firmware version */
 	if (!fu_elantp_hid_device_read_cmd (self,
 					    self->pattern == 0 ? ETP_CMD_I2C_IAP_VERSION : ETP_CMD_I2C_IAP_VERSION_2,
 					    buf, sizeof(buf), error)) {
 		g_prefix_error (error, "failed to read bootloader version: ");
 		return FALSE;
 	}
```
Comment (jinglewu) [inline plugins/elantp/fu-elantp-hid-device.c:177]: Is it possible to change "fwver" to "iap_ver" for fear of confusion with the real TP Firmware version.

File: plugins/elantp/fu-elantp-hid-device.c [hunk 5]
```diff
@@ -143,6 +144,17 @@ fu_elantp_hid_device_setup (FuDevice *device, GError **error)
...
+	if (!fu_elantp_hid_device_read_cmd (self,
+					    ETP_CMD_I2C_GET_HID_ID,
+					    buf, sizeof(buf), error)) {
+		g_prefix_error (error, "failed to read HID ID: ");
+		return FALSE;
+	}
+	tmp = fu_common_read_uint16 (buf, G_LITTLE_ENDIAN);
+	if (tmp != 0xffff)
+		self->pattern = (tmp & 0xff00) >> 8;
+
```
Comment (jinglewu) [inline plugins/elantp/fu-elantp-hid-device.c:159]: if(tmp==0xffff) then -> self->pattern = 0;

File: plugins/elantp/fu-elantp-hid-device.c [hunk 6]
```diff
@@ -308,11 +317,11 @@ fu_elantp_hid_device_write_firmware (FuDevice *device,
...
-	chunks = fu_chunk_array_new (buf + iap_addr, bufsz - iap_addr, 0x0, 0x0, FW_PAGE_SIZE);
+	chunks = fu_chunk_array_new (buf + iap_addr, bufsz - iap_addr, 0x0, 0x0, self->fw_page_size);
 	for (guint i = 0; i < chunks->len; i++) {
 		FuChunk *chk = g_ptr_array_index (chunks, i);
 		guint16 csum_tmp = fu_elantp_calc_checksum (chk->data, chk->data_sz);
-		guint8 blk[FW_PAGE_SIZE + 3];
+		g_autofree guint8 *blk = g_malloc0 (self->fw_page_size + 3);
 
 		/* write block */
 		blk[0] = 0x0B; /* report ID */
```
Comment (jinglewu) [inline plugins/elantp/fu-elantp-hid-device.c:327]: g_usleep (ELANTP_DELAY_WRITE_BLOCK * 1000);

-> msleep(fw_page_size == 512 ? 50 : 35);

File: plugins/elantp/elantp.quirk (hunk 1 above)
Comment (hughsie) [reply plugins/elantp/elantp.quirk:59]: What about the others? Thanks.

File: plugins/elantp/fu-elantp-hid-device.c (hunk 4 above)
Comment (hughsie) [reply plugins/elantp/fu-elantp-hid-device.c:177]: Good idea, fixed.

File: plugins/elantp/fu-elantp-hid-device.c (hunk 5 above)
Comment (hughsie) [reply plugins/elantp/fu-elantp-hid-device.c:159]: 0 is the default, but i'll add it to _init() to make it explicit.

File: plugins/elantp/fu-elantp-hid-device.c (hunk 6 above)
Comment (hughsie) [reply plugins/elantp/fu-elantp-hid-device.c:327]: Thanks, fixed.

File: plugins/elantp/fu-elantp-hid-device.c (hunk 2 above)
Comment (hughsie) [reply plugins/elantp/fu-elantp-hid-device.c:219]: Yup, fixed.

File: plugins/elantp/fu-elantp-hid-device.c (hunk 3 above)
Comment (hughsie) [reply plugins/elantp/fu-elantp-hid-device.c:223]: Fixed, thanks.

File: plugins/elantp/elantp.quirk (hunk 1 above)
Comment (hughsie) [reply plugins/elantp/elantp.quirk:70]: Ahh, the same. Gotcha.

**Comment by hughsie** [comment]
@jinglewu -- looks okay to you now?

**Pull request merged**